src/base/DeadCodeDetector.ml

```diff
@@ -0,0 +1,448 @@
+        | SLiteral.Msg msg ->
+          let new_adts =
+            List.fold_left msg ~init:[] ~f:(fun acc (_, ty, lits') ->
+                iden_of_ADT [ ty ] @ adts_of_literal [ lits' ] @ acc) in
+            new_adts @ res_adts 
+        | Map ((ty1, ty2), _) ->
+            iden_of_ADT [ ty1 ] @ iden_of_ADT [ ty2 ] @ res_adts
+        | ADTValue (_, tys, ts) ->
+            iden_of_ADT tys @ adts_of_literal ts @ res_adts
+        | _ -> [])
+
+  let dedup_id_list = SCIdentifier.dedup_id_list
+
+  let dedup_name_list l =
+    List.dedup_and_sort ~compare:SCIdentifier.Name.compare l
+
+  (* Procedure are checked for seperately due to the discrepency between ER.rep and SR.rep
+     doesn't allow them to be stored as a free variable with the other identifiers. Fix?
+  *)
+  let proc_dict = ref []
+
+  (*
+     Since code is already none pure (proc_dict is a reference), we mark whether fields are read or write
+     in the following, simple, manner
+  *)
+  let read_field = ref []
+
+  let write_field = ref []
+
+  (* Detect Dead Code in a cmod *)
+  let dc_cmod (cmod : cmodule) (elibs : libtree list) =
+    (* Marking a field is used for read or write *)
+    let cfields = List.map cmod.contr.cfields ~f:(fun (f', _, _) -> f') in
+    let mark_field_read f =
+      if SCIdentifier.is_mem_id f cfields then read_field := f :: !read_field
+    in
+    let mark_field_write f =
+      if SCIdentifier.is_mem_id f cfields then write_field := f :: !write_field
+    in
+
+    (******** Checking for dead expressions ********)
+    (* Returns free variables and defined ADT types *)
+    let rec expr_iter (expr, _) =
+      match expr with
+      | Literal l ->
+          let free_tys = adts_of_literal [ l ] in
+          ([], dedup_name_list free_tys)
+      | Var v -> ([ v ], [])
+      | TApp (v, tys) -> ([ v ], iden_of_ADT tys)
+      | Message mlist ->
+          let fvars =
+            List.filter_map mlist ~f:(fun (_, pl) ->
+                match pl with MVar v -> Some v | MLit _ -> None)
+          in
+          (fvars, [])
+      | App (f, actuals) -> (f :: actuals, [])
+      | Constr (_, tys, actuals) | Builtin (_, tys, actuals) ->
+          (actuals, iden_of_ADT tys)
+      | Fixpoint (a, ty, e) | Fun (a, ty, e) ->
+          let e_fv, e_adts = expr_iter e in
+          let e_fv_no_a =
+            List.filter ~f:(fun i -> not @@ SCIdentifier.equal i a) e_fv
+          in
+          (e_fv_no_a, dedup_name_list @@ iden_of_ADT [ ty ] @ e_adts)
+      | Let (i, ty_o, lhs, rhs) ->
+          let fv_rhs, adts_rhs = expr_iter rhs in
+          let fvrhs_no_i =
+            List.filter ~f:(fun x -> not @@ SCIdentifier.equal i x) fv_rhs
+          in
+          if List.mem fv_rhs i ~equal:SCIdentifier.equal then
+            (* LHS is not dead *)
+            let fvlhs, tylhs = expr_iter lhs in
+            let fv = dedup_id_list (fvlhs @ fvrhs_no_i) in
+            let adts = dedup_name_list (adts_rhs @ tylhs) in
+            match ty_o with
```

Comment:
It's ok to not handle the type annotation `ty_o`. No new type can come here that isn't handled elsewhere.